docker/Dockerfile

```diff
@@ -0,0 +1,21 @@
+FROM alpine:3.6
+
+MAINTAINER Ian Denhardt (@zenhack)
+
+ARG SIMP_LE_VERSION=master
+
+ENV PATH=/usr/local/bin:$PATH \
+    SIMP_LE_VERSION=${SIMP_LE_VERSION}
+
+RUN apk --no-cache add bash openssl python3 && \
+    apk --no-cache --virtual .build-deps add gcc git libffi-dev musl-dev openssl-dev py-pip python3-dev && \
+    pip3 install --no-cache-dir --upgrade pip setuptools wheel && \
```

Comment:
Is this line actually needed on alpine? We reinstall/upgrade pip/setuptools/wheel in the venv.sh script because it's supposed to work on a range of distributions, including really stuff like rhel/centos 6.x which has outdated versions of these tools, but it isn't needed on most systems -- part of my gripe about the two Dockerfiles is that there's a certain amount of common boilerplate, getting rid of this would reduce that somewhat.

Also, could we encapsulate the bulk of this RUN command in a shell script, and call it from both Dockerfiles?

I've decided I'm okay having more than one, but I'd still like to factor out the common bits where we can.

Reply:
I don't think the line is needed, on alpine 3.6 the only package pip actually update is wheel, from 28.8 to 29, I'll remove it.

I'm not convinced by the need to factor the common RUN commands between the two.

We'd be replacing this:

```
RUN apk --no-cache add bash openssl python3 && \
    apk --no-cache --virtual .build-deps add gcc git libffi-dev musl-dev openssl-dev py-pip python3-dev && \
    pip3 install --no-cache-dir git+https://github.com/zenhack/simp_le.git@$SIMP_LE_VERSION && \
    apk --update --purge del .build-deps
```

by thoses:

```
COPY .dependencies.sh /simp_le/dependencies.sh

RUN /simp_le/dependencies.sh install && \
    pip3 install --no-cache-dir git+https://github.com/zenhack/simp_le.git@$SIMP_LE_VERSION && \
    /simp_le/dependencies.sh purge && \
    rm /simp_le/dependencies.sh
```
```
#!/bin/bash

case "$1" in
    install)
    apk --no-cache add bash openssl python3 && \
    apk --no-cache --virtual .build-deps add gcc git libffi-dev musl-dev openssl-dev py-pip python3-dev
    ;;
    purge)
    apk --update --purge del .build-deps
    ;;
esac
```

Adding 12 more lines to factor out six (3 line times two Dockerfiles) does not look like a good deal to me. And as a side effect the resulting Dockerfile are really obfuscated.

Reply:
I'd been thinking something more like:

```sh
set -e
apk --no-cache add bash openssl python3
apk --no-cache --virtual .build-deps add gcc git libffi-dev musl-dev openssl-dev py-pip python3-dev
pip3 install --no-cache-dir "$1"
[ -d "$1" ] && rm -rf "$1"
apk --update --purge del .build-deps
```
```
COPY .docker-deps.sh /simp_le/deps.sh
RUN /simp_le/deps.sh git+https://github.com/zenhack/simp_le.git@$SIMP_LE_VERSION
```

```
COPY .docker-deps.sh /simp_le/deps.sh
RUN /simp_le/deps.sh /simp_le/src
```

Thoughts?

Reply:
I've got both Dockerfiles working with that solution but I really have issues with the resulting obfucated Dockerfile that will appear on Dockerhub. You won't have any idea of what it is doing unless you go check the install script on Github.

[Judging](https://github.com/docker-library/python/blob/master/3.6/alpine3.6/Dockerfile) [from](https://github.com/docker-library/mariadb/blob/master/10.3/Dockerfile) [official](https://github.com/docker-library/wordpress/blob/master/php7.1/apache/Dockerfile) [images](https://github.com/docker-library/redis/blob/ebde981d2737c5a618481f766d253afff13aeb9f/3.2/Dockerfile),  from what I know of Docker and how I used/learned it so far, I really don't think hiding Dockerfile's RUN commands behind wrapper script is considered a good practice. On the contrary it seems common practice to have them be extremely clear and explicit about what they do, at the expanse of redundancy. I completely understand the drive to get rid of redundant code, but I think in this case it's missing the point.

If you really aren't ok with not factoring out the common code between the two, I'd prefer to remove the local build Dockerfile and find another way to let people do this (detailing it in the doc maybe ?).

Sorry if you feel I'm being annoying on this one.

Reply:
No, it's fine, I think it's worth hashing stuff like this out -- I am not annoyed in the least.

I'm willing to buy the argument "this is what idiomatic docker looks like."

My own feeling is the shell script I presented isn't terribly obscure/complicated, and the dockerhub links to the github repo, so it would be easy enough to just look at that file. But I'm also coming from a professional software engineering background, maybe I'm under-estimating how hard it would be to follow for someone who has more of an ops background, which I imagine is a decent chunk of docker's user base these days.

In any case, I'm okay leaving this duplicated. Make the other changes we talked about and I'll merge.

Thanks again for the patch!